### PR TITLE
Support Parallel Upgrades

### PR DESCRIPTION
**Pull request by ball-hayden**
BREAKING CHANGE: New API for Upgrades

Defines a new API for upgrades which allows multiple upgrades to be run at once.

---------------------

Self Review:

* [x] Appropriate test coverage
* [x] Relevant Documentation updated

Smoke Tests:

* [x] Tested within PlayerData app context on iOS and Android
* [x] Tested with example app on iOS

### CODE REVIEW

File: android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpdate.kt [hunk 1]
```diff
@@ -20,23 +21,28 @@ val UpgradeModes = mapOf(
 )
 
 class DeviceUpdate(
```
Comment (ball-hayden) [inline android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpdate.kt:23]: For consistency, please can we rename this to `DeviceUpgrade`?

File: android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpdate.kt [hunk 2]
```diff
@@ -61,50 +67,45 @@ class DeviceUpdate(
...
     }
 
     override fun onUpgradeStarted(controller: FirmwareUpgradeController) {}
 
     override fun onStateChanged(prevState: FirmwareUpgradeManager.State, newState: FirmwareUpgradeManager.State) {
         val stateMap = Arguments.createMap()
-        stateMap.putString("bleId", device.address)
+        stateMap.putString("id", id)
         stateMap.putString("state", newState.name)
         manager.updateStateCB(stateMap)
```
Comment (ball-hayden) [inline android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpdate.kt:84]: N.B this should be `upgradeStateChanged` now

```suggestion
        manager.upgradeStateCB(stateMap)
```

File: src/Upgrade.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * McuManager firmware upgrades can actually be performed in few different ways.
+   * These different upgrade modes determine the commands sent after the upload step.
+   *
+   * @see UpgradeMode
+   */
+  upgradeMode?: UpgradeMode;
+}
+
+export type FirmwareUploadState =
```
Comment (ball-hayden) [inline src/Upgrade.ts:46]:
```suggestion
export type FirmwareUpgradeState =
```

Comment (ball-hayden) [reply src/Upgrade.ts:46]: Might be nice to make this an enum for better typescript support.

**Comment by saty9** [comment]
broken for factory test as factory test edges id == bleId when discovered so wrong data is being requested from api

android crashes if an in progress upgrade is cancelled. I think this may be due to a promise being rejected twice so might need to put something in place to catch or prevent that from happenning

**Comment by ball-hayden** [comment]
@saty9 that looks like that comment is for our app, rather than this PR?

**Review by casserlyprogramming — APPROVED** [review]
Looks ok to me. Would be interested in why we are moving away from bleId to device.id but I'm sure there's a good reason.

**Pull request merged**